Refactor planners, memory config, and function extensions

lemillermicrosoft opened the pull request:
This commit includes several updates and refactors to planners,
SemanticMemoryConfig, and function extension classes. Changes include
updating SequentialPlannerConfig to use SemanticMemory, refactoring
tests to use async methods, renaming and updating test cases, and
improving planner configurations. Additionally, FunctionViewExtensions
has been added and refactored, along with updates to method signatures
and code organization. The StepwisePlanner has also been refactored,
and PlannerConfigBase has been updated to improve memory usage and
function filtering.

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

lemillermicrosoft committed:
.Net: Add new token counter implementations to TextChunker (#2840)

Implement MicrosoftML and DeepDev token counters in the TextChunker
example. Update the project file with new package references and modify
the RunExampleWithCustomTokenCounter method to support different token
counter types.

Inspired by #2809 

Fixes #478 

| Iteration | MicrosoftML (ms) | MicrosoftMLRoberta (ms) | SharpToken
(ms) | DeepDev (ms) |

|------------|-------------------|--------------------------|-----------------|--------------|
| 1 | 38 | 10,189 | 14,305 | 16,701 |
| 2 | 36 | 5,581 | 8,381 | 14,214 |
| 3 | 13 | 5,354 | 7,955 | 13,630 |
| 4 | 27 | 5,679 | 9,156 | 16,164 |
| 5 | 16 | 5,158 | 8,657 | 17,276 |
| Average | 26.0 | 7,512.2 | 9,710.8 | 15,597 |




### SharpToken
<sup style="font-size: smaller;">(Avg. Execution Time: 9,710.8 ms)</sup>
```
The city of Venice, located in the northeastern part of Italy,
is renowned for its unique geographical features. Built on more than 100 small islands in a lagoon in the
Adriatic Sea, it has no roads, just canals including the Grand Canal thoroughfare lined with Renaissance and
Gothic palaces. The central square, Piazza San Marco, contains St. Mark's Basilica, which is tiled with Byzantine
mosaics, and the Campanile bell tower offering views of the city's red roofs.
------------------------
The Amazon Rainforest, also known as Amazonia, is a moist broadleaf tropical rainforest in the Amazon biome that
covers most of the Amazon basin of South America. This basin encompasses 7 million square kilometers, of which
5.5 million square kilometers are covered by the rainforest. This region includes territory belonging to nine nations
and 3.4 million square kilometers of uncontacted tribes. The Amazon represents over half of the planet's remaining
rainforests and comprises the largest and most biodiverse tract of tropical rainforest in the world.
------------------------
The Great Barrier Reef is the world's largest coral reef system composed of over 2,900 individual reefs and 900 islands
stretching for over 2,300 kilometers over an area of approximately 344,400 square kilometers. The reef is located in the
Coral Sea, off the coast of Queensland, Australia. The Great Barrier Reef can be seen from outer space and is the world's
biggest single structure made by living organisms. This reef structure is composed of and built by billions of tiny organisms,
known as coral polyps.
```

### MicrosoftML
<sup style="font-size: smaller;">(Avg. Execution Time: 26.0 ms)</sup>
```

The city of Venice,
located in the northeastern part of Italy,
is renowned for its unique
geographical features.
Built on more than 100 small
------------------------
islands in a lagoon in the
Adriatic Sea, it has no roads,
just canals including the Grand Canal
thoroughfare lined with Renaissance and
------------------------
Gothic palaces.
The central square,
Piazza San Marco, contains St.
Mark's Basilica, which is tiled with Byzantine
mosaics,
------------------------
and the Campanile bell tower offering
views of the city's red roofs.
The Amazon Rainforest, also known as Amazonia,
------------------------
is a moist broadleaf tropical
rainforest in the Amazon biome that
covers most of the Amazon
basin of South America.
This basin encompasses 7
------------------------
million square kilometers,
of which
5.
5 million square kilometers
are covered by the rainforest.
This region includes territory
------------------------
belonging to nine nations
and 3.
4 million square kilometers
of uncontacted tribes.
The Amazon represents over
------------------------
half of the planet's remaining
rainforests and comprises the largest and most
biodiverse tract of tropical
rainforest in the world.
------------------------
The Great Barrier Reef is the world's
largest coral reef system composed of over 2,
900 individual reefs and 900 islands
------------------------
stretching for over 2,
300 kilometers over an
area of approximately 344,
400 square kilometers.
The reef is located in the
------------------------
Coral Sea, off the coast of Queensland,
Australia.
The Great Barrier Reef can be seen
from outer space and is the world's
------------------------
biggest single structure
made by living organisms.
This reef structure is composed of and
built by billions of tiny organisms, known as coral polyps.
```

### MicrosoftMLRoberta
<sup style="font-size: smaller;">(Avg. Execution Time: 7,512.2 ms)</sup>
```
The city of Venice, located in the northeastern part of Italy,
is renowned for its unique geographical features. Built on more than 100 small islands in a lagoon in the
Adriatic Sea, it has no roads, just canals including the Grand Canal thoroughfare lined with Renaissance and
Gothic palaces. The central square, Piazza San Marco, contains St. Mark's Basilica, which is tiled with Byzantine
mosaics, and the Campanile bell tower offering views of the city's red roofs.
------------------------
The Amazon Rainforest, also known as Amazonia, is a moist broadleaf tropical rainforest in the Amazon biome that
covers most of the Amazon basin of South America. This basin encompasses 7 million square kilometers, of which
5.5 million square kilometers are covered by the rainforest. This region includes territory belonging to nine nations
and 3.4 million square kilometers of uncontacted tribes. The Amazon represents over half of the planet's remaining
rainforests and comprises the largest and most biodiverse tract of tropical rainforest in the world.
------------------------
The Great Barrier Reef is the world's largest coral reef system composed of over 2,900 individual reefs and 900 islands
stretching for over 2,300 kilometers over an area of approximately 344,400 square kilometers. The reef is located in the
Coral Sea, off the coast of Queensland, Australia. The Great Barrier Reef can be seen from outer space and is the world's
biggest single structure made by living organisms. This reef structure is composed of and built by billions of tiny organisms,
known as coral polyps.
```

### DeepDev
<sup style="font-size: smaller;">(Avg. Execution Time: 15,597 ms)</sup>
```
The city of Venice, located in the northeastern part of Italy,
is renowned for its unique geographical features. Built on more than 100 small islands in a lagoon in the
Adriatic Sea, it has no roads, just canals including the Grand Canal thoroughfare lined with Renaissance and
Gothic palaces. The central square, Piazza San Marco, contains St. Mark's Basilica, which is tiled with Byzantine
mosaics, and the Campanile bell tower offering views of the city's red roofs.
------------------------
The Amazon Rainforest, also known as Amazonia, is a moist broadleaf tropical rainforest in the Amazon biome that
covers most of the Amazon basin of South America. This basin encompasses 7 million square kilometers, of which
5.5 million square kilometers are covered by the rainforest. This region includes territory belonging to nine nations
and 3.4 million square kilometers of uncontacted tribes. The Amazon represents over half of the planet's remaining
rainforests and comprises the largest and most biodiverse tract of tropical rainforest in the world.
------------------------
The Great Barrier Reef is the world's largest coral reef system composed of over 2,900 individual reefs and 900 islands
stretching for over 2,300 kilometers over an area of approximately 344,400 square kilometers. The reef is located in the
Coral Sea, off the coast of Queensland, Australia. The Great Barrier Reef can be seen from outer space and is the world's
biggest single structure made by living organisms. This reef structure is composed of and built by billions of tiny organisms,
known as coral polyps.
```

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

lemillermicrosoft committed:
.Net: Refactor PlannerConfig classes for better organization (#2912)

Moved common properties and methods from SequentialPlannerConfig and
StepwisePlannerConfig to the base class PlannerConfigBase. This change
improves code organization and reduces redundancy. Additionally, added a
new GetSkillFunction property to the PlannerConfigBase class, allowing
for custom function lookup behavior. This change provides flexibility in
how skill functions are retrieved and used within the planning process.

Resolves #2911 

### Motivation and Context

<!-- Thank you for your contribution to the semantic-kernel repo!
Please help reviewers and future users, providing the following
information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->

### Description

<!-- Describe your changes, the overall approach, the underlying design.
These notes will help understanding how your code works. Thanks! -->

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

lemillermicrosoft committed:
Squashed commit of the following:

commit 16fead623a55f75a1834e1bed2d92cf924fd4e67
Author: Lee Miller <lemiller@microsoft.com>
Date:   Fri Sep 22 15:35:26 2023 -0700

    🔄 Refactor planner tests and update function callback handling

    This commit refactors and extends unit tests for various planner
    configurations, including ActionPlannerConfig, StepwisePlannerConfig,
    and SequentialPlannerConfig. It updates the planner tests to use async
    methods and renames some test classes for better organization. The
    function callback handling in SequentialPlanParserTests is also
    modified by replacing calls to SequentialPlanParser.GetFunctionCallback
    with kernel.Functions.GetFunctionCallback() for improved code
    readability and consistency.

commit de79a9eb26b460d1f7a9a461eb60e2103725c91d
Author: Lee Miller <lemiller@microsoft.com>
Date:   Fri Sep 22 14:44:19 2023 -0700

    🔄 Refactor ActionPlanner and improve function handling

    This commit refactors the ActionPlanner class, updating the function
    handling methods and their parameters. It also improves the logging
    mechanism and adds new methods for getting available functions based
    on the semantic query and excluded plugins. The changes improve code readability and
    maintainability. Additionally, unused code has been removed, and new extension
    methods for FunctionView and IReadOnlyFunctionCollection classes have been added,
    making it easier to work with planners. The StepwisePlanner class has been
    simplified by removing unused methods and streamlining the function invocation
    process. A new FunctionUtils class has been added for handling plugin function names.

commit 4ed2bdf6f6379dca0a912bf6ebef0dee3c3349ee
Author: Lee Miller <lemiller@microsoft.com>
Date:   Fri Sep 22 13:16:40 2023 -0700

    🔄 Refactor Semantic Memory Configuration in Planner

    Refactor PlannerConfigBase to include a new SemanticMemoryConfig class,
    which encapsulates semantic memory-related properties. Update
    StepwisePlanner and SequentialPlannerTests to use the new configuration
    structure. This change affects the planner's behavior when searching for
    relevant functions and improves code organization.

commit fe29443ecc61115b967951c3cd525baed2140af4
Author: Lee Miller <lemiller@microsoft.com>
Date:   Thu Sep 21 17:37:07 2023 -0700

    🔍 Enhance StepwisePlanner with function filtering and memory optimization

    This commit improves the StepwisePlanner by introducing filtering and
    searching capabilities for functions based on a given question, and
    optimizing function lookup using semantic memory. It adds a method to
    generate an embedding string for a function, updates the PlannerConfigBase
    with optional semantic memory, and saves available functions to memory for
    better plan creation.

commit e74d8aa1a99cc309b35ffa94b8f4c1fb5936ff57
Author: Lee Miller <lemiller@microsoft.com>
Date:   Thu Sep 21 10:00:56 2023 -0700

    🔄 Update IncludedFunctions type and usage in SKContextExtensions

    Update the type of IncludedFunctions in PlannerConfigBase from
    HashSet<string> to HashSet<(string, string)> to store plugin and function
    names as tuples. Adjust the usage of IncludedFunctions in
    SKContextExtensionsTests and SKContextSequentialPlannerExtensions to
    reflect the new type.

commit ec1271746d55e6d1470b553d66a076bf60d24870
Author: Lee Miller <lemiller@microsoft.com>
Date:   Thu Sep 21 09:34:29 2023 -0700

    🔄 Refactor GetPluginFunction to GetFunctionCallback

    Replace GetPluginFunction with GetFunctionCallback in SequentialPlanner
    and StepwisePlanner. Update SequentialPlanParserTests to use the new
    callback. Rename PlannerConfigBase property accordingly.

commit b06b9be42f8fd2c2d03e2c791fe1c236204f9a77
Author: Lee Miller <lemiller@microsoft.com>
Date:   Wed Sep 20 20:26:56 2023 -0700

    formatting

commit 90aa35e0ecd9d2db8d58d75aab50095dd4d6dc1b
Author: Lee Miller <lemiller@microsoft.com>
Date:   Wed Sep 20 20:22:19 2023 -0700

    🔄 Refactor PlannerConfig classes for better organization

    Moved common properties and methods from SequentialPlannerConfig and
    StepwisePlannerConfig to the base class PlannerConfigBase. This change
    improves code organization and reduces redundancy. Additionally, added a
    new GetSkillFunction property to the PlannerConfigBase class, allowing
    for custom function lookup behavior. This change provides flexibility in
    how skill functions are retrieved and used within the planning process.

commit 83582bb9f59690ff308bbf3b7febadec60aad713
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 11:30:20 2023 -0700

    formatting changes

commit 92b74f4c9da0d4ce4881d05c8bf627bbc0d06d3f
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 11:29:35 2023 -0700

    fix publish mapping

commit 926925c4e993c53d935eb60b57c3f016e4b418b8
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 11:03:44 2023 -0700

    fix project ref

commit 98cffded25474a7267d148fc59d53fd2c8b1c01e
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 11:01:46 2023 -0700

    fix format issues

commit 1a28f3b2dfe86b08330c139433e08937e3ec6d63
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 10:57:56 2023 -0700

    fix reference

commit 407616b73d1ead82fa4f08670187ab56f6e63e4a
Merge: 2f1767f6 544b6c12
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 10:51:39 2023 -0700

    merge main

commit 2f1767f6cfef11666b9a854b4c764e528cd0d35d
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 09:48:58 2023 -0700

    update unit test ns's

commit 6af81d71fc9b9a6e8dbf1d0cb4d01a430924b996
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 09:43:17 2023 -0700

    fix build errors

commit f04744e8b523df79827d8c7252f3e50373525616
Merge: 3666f513 4636b24f
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 09:28:26 2023 -0700

    Merge branch 'merge-planner-packages2' of github.com:hario90/semantic-kernel into merge-planner-packages2

commit 3666f5134e622a86f7b5075b8b989ad01269e3ab
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 09:28:12 2023 -0700

    update nuget description

commit 525482960b1e7ce641ae77a3f0c69cc83be61447
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Fri Sep 22 09:17:21 2023 -0700

    make public classes use Microsoft.SemanticKernel.Planners ns

commit 4636b24f786fdcdb167ceab4d54e5c60876ee70e
Merge: d131532d e3868df2
Author: Mark Wallace <127216156+markwallace-microsoft@users.noreply.github.com>
Date:   Fri Sep 22 07:36:31 2023 +0100

    Merge branch 'main' into merge-planner-packages2

commit d131532d099499101c85be983a3eeca2adba05f7
Merge: 7527a212 283a0dee
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 16:36:45 2023 -0700

    Merge branch 'main' into merge-planner-packages2

commit 7527a2122210c674ec4e367de6b663dc108daf77
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 15:31:20 2023 -0700

    actually delete projects

commit 8acc7805cf1d565273b5905a5d0e8f300ef7a3b8
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 15:28:50 2023 -0700

    remove stepwiseplanner unit test project

commit c66aa8eb4830e45a313ffb32371a57a0c8cbfe6a
Merge: 74a42de0 5e4f08d2
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 13:44:19 2023 -0700

    Merge branch 'main' into merge-planner-packages2

commit 74a42de04a61c514ac8dfb627b093a7f8ff0bd92
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 13:30:47 2023 -0700

    fix warnings

commit 432635c642fa5334552662f410b81395e0a32dae
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 13:19:52 2023 -0700

    remove old projects

commit 6ae2c3fdd4e305ee169bbaff3fcd82a15d63fa94
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 13:19:36 2023 -0700

    update meta package

commit f1d184a9c4140816155ca91f4e70fcba72bc7b7e
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 13:19:13 2023 -0700

    change project referrence

commit fae5323cab1ee20397bb4860af557df65eefc31a
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 13:16:26 2023 -0700

    change references to new project

commit 695d4dd348dada8ad4b523fc7399c160c9296215
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 12:45:13 2023 -0700

    copy planner files to new project

commit b8b24b68c874e13ce05fe17fbc25ce740857b73b
Author: Lisa Harrylock <lisaharrylock@gmail.com>
Date:   Thu Sep 21 11:00:01 2023 -0700

    add Planners.Core and Planners.Core.UnitTests projects

lemillermicrosoft commented:
/sk generate-pr-description

lemillermicrosoft committed:
Squashed commit of the following:

commit 2df149929738b5e88eeda71686c5cb7989735a9d
Author: Lee Miller <lemiller@microsoft.com>
Date:   Fri Sep 22 15:39:56 2023 -0700

    .Net: Refactor PlannerConfig classes for better organization (#2912)

    Moved common properties and methods from SequentialPlannerConfig and
    StepwisePlannerConfig to the base class PlannerConfigBase. This change
    improves code organization and reduces redundancy. Additionally, added a
    new GetSkillFunction property to the PlannerConfigBase class, allowing
    for custom function lookup behavior. This change provides flexibility in
    how skill functions are retrieved and used within the planning process.

    Resolves #2911

    <!-- Thank you for your contribution to the semantic-kernel repo!
    Please help reviewers and future users, providing the following
    information:
      1. Why is this change required?
      2. What problem does it solve?
      3. What scenario does it contribute to?
      4. If it fixes an open issue, please link to the issue here.
    -->

    <!-- Describe your changes, the overall approach, the underlying design.
    These notes will help understanding how your code works. Thanks! -->

    <!-- Before submitting this PR, please make sure: -->

    - [x] The code builds clean without any errors or warnings
    - [x] The PR follows the [SK Contribution
    Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
    and the [pre-submission formatting
    script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
    raises no violations
    - [x] All unit tests pass, and I have added new tests where possible
    - [x] I didn't break anyone :smile:

commit 4a2cf70c9fb254b73b9fac1f99493f62650ce7fa
Author: Lee Miller <lemiller@microsoft.com>
Date:   Fri Sep 22 12:07:27 2023 -0700

    .Net: Add new token counter implementations to TextChunker (#2840)

    Implement MicrosoftML and DeepDev token counters in the TextChunker
    example. Update the project file with new package references and modify
    the RunExampleWithCustomTokenCounter method to support different token
    counter types.

    Inspired by #2809

    Fixes #478

    | Iteration | MicrosoftML (ms) | MicrosoftMLRoberta (ms) | SharpToken
    (ms) | DeepDev (ms) |

    |------------|-------------------|--------------------------|-----------------|--------------|
    | 1 | 38 | 10,189 | 14,305 | 16,701 |
    | 2 | 36 | 5,581 | 8,381 | 14,214 |
    | 3 | 13 | 5,354 | 7,955 | 13,630 |
    | 4 | 27 | 5,679 | 9,156 | 16,164 |
    | 5 | 16 | 5,158 | 8,657 | 17,276 |
    | Average | 26.0 | 7,512.2 | 9,710.8 | 15,597 |
    <sup style="font-size: smaller;">(Avg. Execution Time: 9,710.8 ms)</sup>
    ```
    The city of Venice, located in the northeastern part of Italy,
    is renowned for its unique geographical features. Built on more than 100 small islands in a lagoon in the
    Adriatic Sea, it has no roads, just canals including the Grand Canal thoroughfare lined with Renaissance and
    Gothic palaces. The central square, Piazza San Marco, contains St. Mark's Basilica, which is tiled with Byzantine
    mosaics, and the Campanile bell tower offering views of the city's red roofs.
    ------------------------
    The Amazon Rainforest, also known as Amazonia, is a moist broadleaf tropical rainforest in the Amazon biome that
    covers most of the Amazon basin of South America. This basin encompasses 7 million square kilometers, of which
    5.5 million square kilometers are covered by the rainforest. This region includes territory belonging to nine nations
    and 3.4 million square kilometers of uncontacted tribes. The Amazon represents over half of the planet's remaining
    rainforests and comprises the largest and most biodiverse tract of tropical rainforest in the world.
    ------------------------
    The Great Barrier Reef is the world's largest coral reef system composed of over 2,900 individual reefs and 900 islands
    stretching for over 2,300 kilometers over an area of approximately 344,400 square kilometers. The reef is located in the
    Coral Sea, off the coast of Queensland, Australia. The Great Barrier Reef can be seen from outer space and is the world's
    biggest single structure made by living organisms. This reef structure is composed of and built by billions of tiny organisms,
    known as coral polyps.
    ```
    <sup style="font-size: smaller;">(Avg. Execution Time: 26.0 ms)</sup>
    ```

    The city of Venice,
    located in the northeastern part of Italy,
    is renowned for its unique
    geographical features.
    Built on more than 100 small
    ------------------------
    islands in a lagoon in the
    Adriatic Sea, it has no roads,
    just canals including the Grand Canal
    thoroughfare lined with Renaissance and
    ------------------------
    Gothic palaces.
    The central square,
    Piazza San Marco, contains St.
    Mark's Basilica, which is tiled with Byzantine
    mosaics,
    ------------------------
    and the Campanile bell tower offering
    views of the city's red roofs.
    The Amazon Rainforest, also known as Amazonia,
    ------------------------
    is a moist broadleaf tropical
    rainforest in the Amazon biome that
    covers most of the Amazon
    basin of South America.
    This basin encompasses 7
    ------------------------
    million square kilometers,
    of which
    5.
    5 million square kilometers
    are covered by the rainforest.
    This region includes territory
    ------------------------
    belonging to nine nations
    and 3.
    4 million square kilometers
    of uncontacted tribes.
    The Amazon represents over
    ------------------------
    half of the planet's remaining
    rainforests and comprises the largest and most
    biodiverse tract of tropical
    rainforest in the world.
    ------------------------
    The Great Barrier Reef is the world's
    largest coral reef system composed of over 2,
    900 individual reefs and 900 islands
    ------------------------
    stretching for over 2,
    300 kilometers over an
    area of approximately 344,
    400 square kilometers.
    The reef is located in the
    ------------------------
    Coral Sea, off the coast of Queensland,
    Australia.
    The Great Barrier Reef can be seen
    from outer space and is the world's
    ------------------------
    biggest single structure
    made by living organisms.
    This reef structure is composed of and
    built by billions of tiny organisms, known as coral polyps.
    ```
    <sup style="font-size: smaller;">(Avg. Execution Time: 7,512.2 ms)</sup>
    ```
    The city of Venice, located in the northeastern part of Italy,
    is renowned for its unique geographical features. Built on more than 100 small islands in a lagoon in the
    Adriatic Sea, it has no roads, just canals including the Grand Canal thoroughfare lined with Renaissance and
    Gothic palaces. The central square, Piazza San Marco, contains St. Mark's Basilica, which is tiled with Byzantine
    mosaics, and the Campanile bell tower offering views of the city's red roofs.
    ------------------------
    The Amazon Rainforest, also known as Amazonia, is a moist broadleaf tropical rainforest in the Amazon biome that
    covers most of the Amazon basin of South America. This basin encompasses 7 million square kilometers, of which
    5.5 million square kilometers are covered by the rainforest. This region includes territory belonging to nine nations
    and 3.4 million square kilometers of uncontacted tribes. The Amazon represents over half of the planet's remaining
    rainforests and comprises the largest and most biodiverse tract of tropical rainforest in the world.
    ------------------------
    The Great Barrier Reef is the world's largest coral reef system composed of over 2,900 individual reefs and 900 islands
    stretching for over 2,300 kilometers over an area of approximately 344,400 square kilometers. The reef is located in the
    Coral Sea, off the coast of Queensland, Australia. The Great Barrier Reef can be seen from outer space and is the world's
    biggest single structure made by living organisms. This reef structure is composed of and built by billions of tiny organisms,
    known as coral polyps.
    ```
    <sup style="font-size: smaller;">(Avg. Execution Time: 15,597 ms)</sup>
    ```
    The city of Venice, located in the northeastern part of Italy,
    is renowned for its unique geographical features. Built on more than 100 small islands in a lagoon in the
    Adriatic Sea, it has no roads, just canals including the Grand Canal thoroughfare lined with Renaissance and
    Gothic palaces. The central square, Piazza San Marco, contains St. Mark's Basilica, which is tiled with Byzantine
    mosaics, and the Campanile bell tower offering views of the city's red roofs.
    ------------------------
    The Amazon Rainforest, also known as Amazonia, is a moist broadleaf tropical rainforest in the Amazon biome that
    covers most of the Amazon basin of South America. This basin encompasses 7 million square kilometers, of which
    5.5 million square kilometers are covered by the rainforest. This region includes territory belonging to nine nations
    and 3.4 million square kilometers of uncontacted tribes. The Amazon represents over half of the planet's remaining
    rainforests and comprises the largest and most biodiverse tract of tropical rainforest in the world.
    ------------------------
    The Great Barrier Reef is the world's largest coral reef system composed of over 2,900 individual reefs and 900 islands
    stretching for over 2,300 kilometers over an area of approximately 344,400 square kilometers. The reef is located in the
    Coral Sea, off the coast of Queensland, Australia. The Great Barrier Reef can be seen from outer space and is the world's
    biggest single structure made by living organisms. This reef structure is composed of and built by billions of tiny organisms,
    known as coral polyps.
    ```

    <!-- Before submitting this PR, please make sure: -->

    - [ ] The code builds clean without any errors or warnings
    - [ ] The PR follows the [SK Contribution
    Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
    and the [pre-submission formatting
    script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
    raises no violations
    - [ ] All unit tests pass, and I have added new tests where possible
    - [ ] I didn't break anyone :smile:

lemillermicrosoft commented:
/sk generate-pr-description

lemillermicrosoft commented:
/sk generate-pr-description

lemillermicrosoft committed:
small fix

hario90 commented on dotnet/src/Planners/Planners.Core/Extensions/FunctionViewExtensions.cs:
```diff
@@ -0,0 +1,53 @@
+/// <summary>
+/// Provides extension methods for the <see cref="FunctionView"/> class.
+/// </summary>
+public static class FunctionViewExtensions
```
What is the reason for making this public? Is this something we want to expose to planners outside this package?

lemillermicrosoft replied:
Was just being consistent with the other extensions. At one point I made them all internal. But we did get feedback in the past to make things easy to re-use. Either way we should be confident in the final call before going to main.

lemillermicrosoft commented on dotnet/src/Planners/Planners.Core/Utils/FunctionUtils.cs:
```diff
@@ -0,0 +1,11 @@
+
+internal static class FunctionUtils
+{
+    internal static void GetFunctionCallbackNames(string pluginFunctionName, out string pluginName, out string functionName)
```
Did not give naming hee much thought.

hario90 replied:
Suggestion: GetPluginAndFunctionNamesFromPlanData

lemillermicrosoft committed:
cleanup file

hario90 commented on dotnet/src/Planners/Planners.Core/Extensions/ReadOnlyFunctionCollectionPlannerExtensions.cs:
```diff
@@ -109,20 +147,18 @@ public static class SKContextSequentialPlannerExtensions
             {
-                logger ??= context.LoggerFactory.CreateLogger(typeof(SKContext));
-                if (logger.IsEnabled(LogLevel.Debug))
+                if (logger is not null && logger.IsEnabled(LogLevel.Debug))
```
Not important right now but it seems like checking whether the logger is enabled should be the logger's job which would reduce checks like this.

lemillermicrosoft replied:
Yeah agreed -- I think this was related to past feedback from folks like StephenToub though that it's better to do this everywhere to avoid the string allocations when there is no logger and/or it's not enabled for that log level.

lemillermicrosoft replied:
We're not doing that everywhere, but yeah, might be worth re-considering in places.

hario90 commented on dotnet/src/Planners/Planners.Core/Extensions/ReadOnlyFunctionCollectionPlannerExtensions.cs:
```diff
@@ -109,20 +147,18 @@ public static class SKContextSequentialPlannerExtensions
-        SKContext context,
         IEnumerable<FunctionView> availableFunctions,
         IAsyncEnumerable<MemoryQueryResult> memories,
+        ILogger? logger = null,
```
nit: to remove the null check you could use a NullLogger
```suggestion
        ILogger? logger = NullLogger.Instance,
```

lemillermicrosoft committed:
📝 Add missing namespaces in test files

Add Microsoft.SemanticKernel.Planners namespace to
SequentialPlanParserTests.cs and import Memory, Orchestration, and
XunitHelpers namespaces in ReadOnlyFunctionCollectionExtensionsTests.cs.

markwallace-microsoft committed:
.Net: Rename ImportXXXPlugins methods to ImportXXXFunctions for GRPC (#2943)

### Motivation and Context

### Description

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

SergeyMenshykh committed:
.Net: [OpenApi] Pipe delimited query string parameters (#2941)

### Motivation and Context
This change is required to support two ways of serializing array query
string parameters of 'pipeDelimited' style:

A query string parameter per array item - p1=a&p1=b&p1=c
A space-separated value per array item - p1=a|b|c


![image](https://github.com/microsoft/semantic-kernel/assets/68852919/4974647a-fe9c-4561-b4fd-bf380ae6ad4a)

Related issue - https://github.com/microsoft/semantic-kernel/issues/2745

### Description

- The `PipeDelimitedStyleParametersSerializer` class has been added to
perform `pipeDelimited` style parameter serialization. For more details
about parameter styles, please refer to the
[link](https://swagger.io/specification/v3/#parameter-object).
- The `QueryStringBuilder` class has been modified to use the
`PipeDelimitedStyleParametersSerializer` for the `pipeDelimited` style
parameter serialization.

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

markwallace-microsoft committed:
.Net: Rename ImportXXXPlugins methods to ImportXXXFunctions (#2937)

### Motivation and Context

As part of the Skill -> Plugin rename the `ImportSkill` and
`ImportSemanticSkillFromDirectory` where renamed to use the term
`Plugin`. These methods actually import functions to the Kernel so this
PR renames them to match what they actually do.

We want to reserve `ImportPlugin` for use later when we add more plugin
support to the SK.

So now we do this:
`var functions = kernel.ImportFunctionsFromDirectory(...);`

In future we will do this
`var plugin = kernel.ImportPluginFromDirectory(...);`

The plugin instance will contain a list of functions in addition to
other data about the plugin.

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

WeihanLi committed:
.Net: Update PlannerConfigBase.cs (#2954)

### Motivation and Context

To keep code clean

### Description

Remove unnecessary init

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

WeihanLi committed:
.Net: Update PromptTemplate.cs (#2955)

### Motivation and Context

Keep code clean

### Description

let `_params` to be `readonly` since nowhere updates it

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

JadynWong committed:
.Net: Postgres memory store add simple constructor (#2688)

### Motivation and Context

<!-- Thank you for your contribution to the semantic-kernel repo!
Please help reviewers and future users, providing the following
information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->

The Postgres memory store adds a simple constructor using a connection
string. The `NpgsqlDataSource` is managed internally by the postgres
memory store.

### Description

<!-- Describe your changes, the overall approach, the underlying design.
These notes will help understanding how your code works. Thanks! -->

- Similar to the recent changes to the redis memory store. which allows
the user to pass the connection string directly. the `NpgsqlDataSource`
is managed internally.
- Fixed postgres integration test (caused by #2419)

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

---------

Co-authored-by: Lee Miller <lemiller@microsoft.com>
Co-authored-by: Dmytro Struk <13853051+dmytrostruk@users.noreply.github.com>

hario90 committed:
.Net: Merge planner packages2 (#2931)

### Motivation and Context

Combining the planner files into one package makes it easier to provide
a consistent interface for planners.
Fixes #2856

This introduces breaking changes since the namespaces for the planners
changed.

### Description
* Created Planners.Core and Planners.Core.UnitTests projects under new
folder Planners
* Created a folder for each type of planner within the Planners project
* Moved planner files to their new project and folder.
* Removed old projects

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution
Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
and the [pre-submission formatting
script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts)
raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile: - Namespaces for planner files
changed

---------

Co-authored-by: Mark Wallace <127216156+markwallace-microsoft@users.noreply.github.com>

lemillermicrosoft committed:
refactor(Planners.Core): Rename GetFunctionCallbackNames to SplitPluginFunctionName.

Renamed the GetFunctionCallbackNames method in FunctionUtils to SplitPluginFunctionName for better clarity and updated its usage in ActionPlanner, SequentialPlanParser, and StepwisePlanner.

lemillermicrosoft committed:
refactor(FunctionViewExtensions): Change visibility of FunctionViewExtensions and ToManualString method.

Changed the visibility of FunctionViewExtensions class and ToManualString method from public to internal to limit their accessibility outside the assembly.

lemillermicrosoft committed:
refactor(Planners.Core): Use NullLogger.Instance as default logger in GetRelevantFunctionsAsync.

Changed the default value of the logger parameter in GetRelevantFunctionsAsync to NullLogger.Instance, removing the need for null checks when calling logger methods.

lemillermicrosoft committed:
refactor(Planners): Rename SemanticMemory property to SemanticMemoryConfig.

Renamed the SemanticMemory property in PlannerConfigBase and related classes to SemanticMemoryConfig to better reflect its purpose. Updated references and tests accordingly.

lemillermicrosoft committed:
feat(Planners.Core): Add SemanticMemoryConfig class.

Introduce a new SemanticMemoryConfig class to configure semantic memory settings for filtering function lookup during plan creation. This includes properties for IncludedFunctions, Memory, MaxRelevantFunctions, and RelevancyThreshold.

lemillermicrosoft committed:
test(Planners.Core.UnitTests): Remove unused import in ReadOnlyFunctionCollectionExtensionsTests.

Removed an unnecessary import of Microsoft.SemanticKernel.Planning in ReadOnlyFunctionCollectionExtensionsTests.cs.

lemillermicrosoft committed:
refactor(Planners.Core): Remove unnecessary using statements in PlannerConfig classes.

Removed unused using statements in ActionPlannerConfig, ReadOnlyFunctionCollectionPlannerExtensions, PlannerConfigBase, SequentialPlannerConfig, and StepwisePlannerConfig files.
